lib/ecto/multi.ex

```diff
@@ -0,0 +1,364 @@
+defmodule Ecto.Multi do
+  @moduledoc """
+  Ecto.Multi is a data structure that allows grouping multiple Repo operations
+  together.
+
+  Ecto.Multi makes it possible to pack operations that should be performed
+  together (in a single database transaction) and gives a way to introspect
+  the queued operations without actually performing them.
+  Each operation is given a name that is unique and will identify it's result
+  or will help to identify place of failure in case it occurs.
+
+  All operations will be executed in the order they were added.
+
+  The `Ecto.Multi` structure should be considered opaque. You can use
+  `%Ecto.Multi{}` to pattern match the type, but accessing fields or
+  directly modifying them is not advised.
+  `Ecto.Multi.to_list/1` returns a canonical representation of the structure
+  that can be used for introspection.
+
+  ## Changesets
+
+  If Multi contains operations that accept changesets (like `insert/4`,
+  `update/4` or `delete/4`) they will be checked before starting transaction
+  if any of them isn't invalid. In case any has errors, the transaction won't
+  be even started and the error will be immediately returned.
+
+  ## Run
+
+  Multi allows you to run arbitrary functions as part of your transaction via
+  the `run/3` and `run/5`. Those functions will receive changes so far as the
+  first argument and have to return `{:ok, value}` or `{:error, value}` as
+  their result. Returning an error will abort any further operations and
+  make the whole multi fail.
```

Comment:
I think it would be great to include an example of using `run` in the example `Service` module below. What do you think?